### PR TITLE
perf: bicep structure upgrade

### PR DESCRIPTION
**Pull request by wenytang-ms**
now only support scaffold
1. tab app (aad + frontend + simpleauth)
2. fix test case (aad + frontend + simpleauth)

### CODE REVIEW

File: packages/fx-core/src/common/armInterface.ts [hunk 1]
```diff
@@ -37,3 +37,13 @@ export interface ScaffoldArmTemplateResult extends Record<string, unknown> {
...
 }
+
+export interface ArmTemplateResult extends Record<string, unknown> {
+  Provision?: {
+    Orchestration: string;
+    Reference?: Record<string, unknown>;
+    Modules?: { [moduleFileName: string]: string };
+  };
+  Configuration?: { Orchestration: string; Modules?: { [moduleFileName: string]: string } };
+  Parameters?: string;
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/common/armInterface.ts:48]: sure

File: packages/fx-core/src/plugins/resource/bot/plugin.ts [hunk 2]
```diff
@@ -198,82 +198,46 @@ export class TeamsBotImpl {
...
+        await fs.readFile(
+          path.join(bicepTemplateDir, Bicep.ParameterFileName),
+          ConstantString.UTF8Encoding
+        )
+      ),
     };
 
     Logger.info(Messages.SuccessfullyGenerateArmTemplatesBot);
-    return ResultFactory.Success(result);
+    return ResultFactory.Success(result1);
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/bot/plugin.ts:240]: done

File: packages/fx-core/src/plugins/resource/function/constants.ts [hunk 3]
```diff
@@ -181,7 +181,10 @@ export class QuestionValidationFunc {
 }
 
 export class FunctionBicep {
-  static readonly functionEndpoint: string = "functionProvision.outputs.functionEndpoint";
+  static readonly functionEndpoint: string =
+    "provisionOutputs.functionProvision.outputs.functionEndpoint";
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/function/constants.ts:185]: fixed

File: packages/fx-core/src/plugins/solution/fx-solution/arm.ts [hunk 4]
```diff
@@ -429,21 +446,42 @@ async function doGenerateArmTemplate(ctx: SolutionContext): Promise<Result<any,
...
-    const bicepOrchestrationFileContent = bicepOrchestrationTemplate.getOrchestrationFileContent();
+    let bicepOrchestrationProvisionContent = await fs.readFile(
+      path.join(getTemplatesFolder(), "plugins", "solution", "provision.bicep"),
+      ConstantString.UTF8Encoding
+    );
+    let bicepOrchestrationConfigContent = await fs.readFile(
+      path.join(getTemplatesFolder(), "plugins", "solution", "config.bicep"),
+      ConstantString.UTF8Encoding
+    );
+    bicepOrchestrationProvisionContent +=
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/solution/fx-solution/arm.ts:457]: done.

File: packages/fx-core/src/plugins/solution/fx-solution/arm.ts [hunk 5]
```diff
@@ -429,21 +446,42 @@ async function doGenerateArmTemplate(ctx: SolutionContext): Promise<Result<any,
...
+      bicepOrchestrationTemplate.getOrchestractionProvisionContent();
+    bicepOrchestrationConfigContent += bicepOrchestrationTemplate.getOrchestractionConfigContent();
     const templateFolderPath = path.join(ctx.root, templatesFolder);
     await fs.ensureDir(templateFolderPath);
+    await fs.ensureDir(path.join(templateFolderPath, "teamsFxConfiguration"));
+    await fs.ensureDir(path.join(templateFolderPath, "provision"));
+    // generate main.bicep
+    await fs.copyFile(
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/solution/fx-solution/arm.ts:465]: sure, will check file exist before copy

File: packages/fx-core/src/plugins/solution/fx-solution/arm.ts [hunk 6]
```diff
@@ -429,21 +446,42 @@ async function doGenerateArmTemplate(ctx: SolutionContext): Promise<Result<any,
...
+      path.join(getTemplatesFolder(), "plugins", "solution", "main.bicep"),
+      path.join(templateFolderPath, bicepOrchestrationFileName)
+    );
+    // generate provision.bicep
     await fs.writeFile(
-      path.join(templateFolderPath, bicepOrchestrationFileName),
-      bicepOrchestrationFileContent
+      path.join(templateFolderPath, bicepOrchestrationProvisionFileName),
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/solution/fx-solution/arm.ts:471]: updated as appendFile

File: packages/fx-core/src/plugins/solution/fx-solution/arm.ts [hunk 7]
```diff
@@ -429,21 +446,42 @@ async function doGenerateArmTemplate(ctx: SolutionContext): Promise<Result<any,
...
     );
 
     // Output bicep module files from each resource plugin
-    const modulesFolderPath = path.join(templateFolderPath, modulesFolder);
-    await fs.ensureDir(modulesFolderPath);
     for (const module of moduleFiles) {
       // module[0] contains relative path to template folder, e.g. "./modules/frontendHosting.bicep"
-      await fs.writeFile(path.join(templateFolderPath, module[0]), module[1]);
+      const res = bicepOrchestrationTemplate.applyReference(module[1]);
+      await fs.writeFile(path.join(templateFolderPath, module[0]), res);
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/solution/fx-solution/arm.ts:484]: updated.

File: packages/fx-core/src/plugins/solution/fx-solution/arm.ts [hunk 8]
```diff
@@ -541,28 +579,38 @@ export class ArmTemplateRenderContext {
...
+          ProvisionPath: generateBicepModuleProvisionFilePath(moduleFileName),
+        };
+      }
+    }
+
+    if (configs) {
+      for (const module of Object.entries(configs)) {
+        const moduleFileName = module[0];
+        pluginOutputContext.Modules![moduleFileName] = {
+          ConfigPath: generateBicepModuleConfigFilePath(moduleFileName),
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/solution/fx-solution/arm.ts:604]: yes, will update later

File: packages/fx-core/src/plugins/solution/fx-solution/arm.ts [hunk 9]
```diff
@@ -646,22 +711,28 @@ class BicepOrchestrationContent {
...
+  return `./teamsFxConfiguration/${moduleFileName}.bicep`;
 }
 
 function expandParameterPlaceholders(ctx: SolutionContext, parameterContent: string): string {
   const azureSolutionSettings = ctx.projectSettings?.solutionSettings as AzureSolutionSettings;
   const plugins = getActivatedResourcePlugins(azureSolutionSettings); // This function ensures return result won't be empty
   const stateVariables: Record<string, Record<string, any>> = {};
-  const availableVariables: Record<string, Record<string, any>> = { state: stateVariables };
+  const availableVariables: Record<string, Record<string, any>> = {
+    provisionParameters: { value: stateVariables },
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/solution/fx-solution/arm.ts:734]: revert this fix

File: packages/fx-core/templates/plugins/resource/bot/bicep/botConfiguration.template.v2.bicep [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  name: '${botWebAppName}/appsettings'
+  properties: union({
+    INITIATE_LOGIN_ENDPOINT: uri(provisionOutputs.botOutput.value.webAppEndpoint, 'auth-start.html')
+    M365_AUTHORITY_HOST: m365OauthAuthorityHost
+    M365_CLIENT_ID: m365ClientId
+    M365_CLIENT_SECRET: m365ClientSecret
+    M365_TENANT_ID: m365TenantId
+    M365_APPLICATION_ID_URI: m365ApplicationIdUri
+    {{#contains 'fx-resource-function' Plugins}}
+    API_ENDPOINT: provisionOutputs.functionOutput.value.endpoint
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/templates/plugins/resource/bot/bicep/botConfiguration.template.v2.bicep:42]: yes, function is not support now. will udpate later

File: packages/fx-core/src/plugins/resource/identity/index.ts [hunk 11]
```diff
@@ -136,49 +136,63 @@ export class IdentityPlugin implements Plugin {
...
-    const moduleOrchestrationFilePath = path.join(
-      bicepTemplateDirectory,
-      Bicep.ModuleOrchestrationFileName
-    );
-    const outputTemplateFilePath = path.join(
-      bicepTemplateDirectory,
-      Bicep.OutputOrchestrationFileName
+      IdentityBicepFile.moduleTempalteV2Filename
     );
+    // const moduleContentResult = await generateBicepFiles(moduleTemplateFilePath, context);
```
Comment (dilin-MS) [inline packages/fx-core/src/plugins/resource/identity/index.ts:141]: Remove useless code.

File: packages/fx-core/src/plugins/resource/identity/index.ts [hunk 12]
```diff
@@ -136,49 +136,63 @@ export class IdentityPlugin implements Plugin {
...
-        OutputTemplate: {
-          Content: await fs.readFile(outputTemplateFilePath, ConstantString.UTF8Encoding),
+        Modules: {
+          identityProvision: await fs.readFile(moduleTemplateFilePath, ConstantString.UTF8Encoding),
         },
       },
     };
+
+    // const result: ScaffoldArmTemplateResult = {
+    //   Modules: {
```
Comment (dilin-MS) [inline packages/fx-core/src/plugins/resource/identity/index.ts:174]: Same here.

File: packages/fx-core/src/plugins/resource/identity/index.ts (hunk 11 above)
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/identity/index.ts:141]: done

File: packages/fx-core/src/plugins/resource/identity/index.ts (hunk 12 above)
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/identity/index.ts:174]: done

File: packages/fx-core/.fx/subscriptionInfo.json [hunk 13]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (blackchoey) [inline packages/fx-core/.fx/subscriptionInfo.json:1]: remove this file?

File: packages/fx-core/src/plugins/solution/fx-solution/arm.ts [hunk 14]
```diff
@@ -242,7 +241,7 @@ export async function doDeployArmTemplates(ctx: SolutionContext): Promise<Result
       .finally(() => {
         deployCtx.finished = true;
       });
-
+    console.log(result);
```
Comment (blackchoey) [inline packages/fx-core/src/plugins/solution/fx-solution/arm.ts:244]: can we remove it?

File: packages/fx-core/templates/azure/config.bicep [hunk 15]
```diff
@@ -0,0 +1,5 @@
+@secure()
```
Comment (blackchoey) [inline packages/fx-core/templates/azure/config.bicep:1]: are `packages/fx-core/templates/azure/xxx` test files? we may need to add cleanup code for test files, in the meanwhile, we may set project root to somewhere under test folder

File: packages/fx-core/tests/plugins/resource/frontend/unit/generateArmTemplates.test.ts [hunk 16]
```diff
@@ -37,49 +37,50 @@ describe("FrontendGenerateArmTemplates", () => {
...
+              ProvisionPath: `./${testModuleFileName}`,
             },
           },
         },
       },
     };
     chai.assert.isTrue(result.isOk());
     if (result.isOk()) {
-      const expectedResult = mockSolutionUpdateArmTemplates(
+      // const expectedResult = mockSolutionUpdateArmTemplates(
```
Comment (blackchoey) [inline packages/fx-core/tests/plugins/resource/frontend/unit/generateArmTemplates.test.ts:55]: remove the unused comment in all test files?

File: packages/fx-core/.fx/subscriptionInfo.json (hunk 13 above)
Comment (wenytang-ms) [reply packages/fx-core/.fx/subscriptionInfo.json:1]: sorry, will remove this file

File: packages/fx-core/src/plugins/solution/fx-solution/arm.ts (hunk 14 above)
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/solution/fx-solution/arm.ts:244]: thanks

File: packages/fx-core/tests/plugins/resource/frontend/unit/generateArmTemplates.test.ts (hunk 16 above)
Comment (wenytang-ms) [reply packages/fx-core/tests/plugins/resource/frontend/unit/generateArmTemplates.test.ts:55]: sure

File: packages/fx-core/templates/azure/config.bicep (hunk 15 above)
Comment (wenytang-ms) [reply packages/fx-core/templates/azure/config.bicep]: seems solution.scaffold.tests.ts create this test folder, already removed

**Review by blackchoey — APPROVED** [review]

**Pull request closed without merging**